## 1. Environment Setup and Data Loading

### 📝 Task 1.1: Setup and Data Loading

In [5]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers, models

(x_train, y_train), (x_test, y_test) = keras.datasets.fashion_mnist.load_data()

In [6]:
x_train = x_train / 255.0
x_test = x_test / 255.0

In [7]:
x_train_mlp = x_train.reshape((x_train.shape[0], 28*28))
x_test_mlp = x_test.reshape((x_test.shape[0], 28*28))
x_train_mlp.shape

(60000, 784)

In [8]:
x_train_cnn = x_train.reshape((x_train.shape[0], 28, 28, 1))
x_test_cnn = x_test.reshape((x_test.shape[0], 28, 28, 1))
x_train_cnn.shape

(60000, 28, 28, 1)

## 2. Model Implementation

### 📝 Task 2.1: Implement and Compile the MLP Model

In [9]:
mlp_model = models.Sequential([
    layers.Input(shape=(784,)),
    layers.Dense(256, activation='relu'),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

mlp_model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
mlp_model.summary()


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 256)            │       200,960 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 235,146 (918.54 KB)

 Trainable params: 235,146 (918.54 KB)

 Non-trainable params: 0 (0.00 B)

### 📝 Task 2.2: Implement and Compile the CNN Model

In [10]:
cnn_model = models.Sequential([
    layers.Conv2D(16, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D(2,2),
    layers.Conv2D(32, (3,3), activation='relu'),
    layers.MaxPooling2D(2,2),
    layers.Flatten(),
    layers.Dense(64, activation='relu'),
    layers.Dense(10, activation='softmax')
])

cnn_model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])

cnn_model.summary()


c:\Users\HSN\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 16)     │           160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 16)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 32)     │         4,640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 32)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 800)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 64)             │        51,264 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 56,714 (221.54 KB)

 Trainable params: 56,714 (221.54 KB)

 Non-trainable params: 0 (0.00 B)

## 3. Training and Evaluation

### 📝 Task 3.1: Train the MLP

In [11]:
history_mlp = mlp_model.fit(
    x_train_mlp, y_train,
    epochs=5,
    batch_size=64,
    validation_data=(x_test_mlp, y_test),
    verbose=2
)

Epoch 1/5
938/938 - 5s - 5ms/step - accuracy: 0.8250 - loss: 0.4839 - val_accuracy: 0.8542 - val_loss: 0.4030
Epoch 2/5
938/938 - 3s - 4ms/step - accuracy: 0.8677 - loss: 0.3609 - val_accuracy: 0.8530 - val_loss: 0.4191
Epoch 3/5
938/938 - 4s - 4ms/step - accuracy: 0.8793 - loss: 0.3231 - val_accuracy: 0.8705 - val_loss: 0.3589
Epoch 4/5
938/938 - 4s - 4ms/step - accuracy: 0.8881 - loss: 0.2985 - val_accuracy: 0.8686 - val_loss: 0.3627
Epoch 5/5
938/938 - 4s - 4ms/step - accuracy: 0.8950 - loss: 0.2804 - val_accuracy: 0.8810 - val_loss: 0.3368


### 📝 Task 3.2: Train the CNN

In [12]:
history_cnn = cnn_model.fit(
    x_train_cnn, y_train,
    epochs=5,
    batch_size=64,
    validation_data=(x_test_cnn, y_test),
    verbose=2
)


Epoch 1/5
938/938 - 10s - 11ms/step - accuracy: 0.7926 - loss: 0.5746 - val_accuracy: 0.8444 - val_loss: 0.4375
Epoch 2/5
938/938 - 9s - 9ms/step - accuracy: 0.8627 - loss: 0.3776 - val_accuracy: 0.8647 - val_loss: 0.3787
Epoch 3/5
938/938 - 10s - 11ms/step - accuracy: 0.8796 - loss: 0.3324 - val_accuracy: 0.8724 - val_loss: 0.3478
Epoch 4/5
938/938 - 10s - 10ms/step - accuracy: 0.8917 - loss: 0.3014 - val_accuracy: 0.8878 - val_loss: 0.3138
Epoch 5/5
938/938 - 10s - 10ms/step - accuracy: 0.8968 - loss: 0.2831 - val_accuracy: 0.8908 - val_loss: 0.3024


### 📝 Task 3.3: Evaluate and Report

In [13]:
mlp_test_loss, mlp_test_acc = mlp_model.evaluate(x_test_mlp, y_test, verbose=0)
cnn_test_loss, cnn_test_acc = cnn_model.evaluate(x_test_cnn, y_test, verbose=0)

print(f"MLP Test Accuracy: {mlp_test_acc:.4f}, Loss: {mlp_test_loss:.4f}")
print(f"CNN Test Accuracy: {cnn_test_acc:.4f}, Loss: {cnn_test_loss:.4f}")


MLP Test Accuracy: 0.8810, Loss: 0.3368
CNN Test Accuracy: 0.8908, Loss: 0.3024


## 4. Resource Usage Comparison

### 📝 Task 4.1: Count Trainable Parameters

In [14]:
mlp_model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 256)            │       200,960 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 705,440 (2.69 MB)

 Trainable params: 235,146 (918.54 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 470,294 (1.79 MB)

In [15]:
cnn_model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 16)     │           160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 16)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 32)     │         4,640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 32)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 800)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 64)             │        51,264 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │           650 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 170,144 (664.63 KB)

 Trainable params: 56,714 (221.54 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 113,430 (443.09 KB)

In [16]:
mlp_model.save("mlp_model.h5")
cnn_model.save("cnn_model.h5")

In [17]:
import os
mlp_size = os.path.getsize("mlp_model.h5") / (1024 * 1024)
cnn_size = os.path.getsize("cnn_model.h5") / (1024 * 1024)

print(f"MLP Model Size: {mlp_size:.2f} MB")
print(f"CNN Model Size: {cnn_size:.2f} MB")

MLP Model Size: 2.72 MB
CNN Model Size: 0.69 MB


### 📝 Task 4.3: Estimate Computational Resources

| Metric                                               | **MLP Model**            | **CNN Model**            |
| ---------------------------------------------------- | ------------------------ | ------------------------ |
| **Trainable Parameters**                             | 235,146                  | 56,714                   |
| **Optimizer Parameters**                             | 470,294                  | 113,430                  |
| **Total Parameters**                                 | 705,440                  | 170,144                  |
| **Saved Model Size**                                 | 2.72 MB                  | 0.69 MB                  |
| **Estimated FLOPs (Inference)**                      | ≈ 2.3 Million            | ≈ 15–20 Million          |
| **Estimated FLOPs (Training)**                       | ≈ 4.6 Million            | ≈ 30–40 Million          |
| **Training Memory (Params + Gradients + Optimizer)** | ≈ 3× model size ≈ 8–9 MB | ≈ 3× model size ≈ 2–3 MB |


## 5. Final Report and Conclusio

### 📝 Task 5.1: Write the Conclusion

| Model   | Test Accuracy | Test Loss | Trainable Parameters | Saved Model Size (MB) | FLOPs (Training) | FLOPs (Inference) | Training Memory (MB) |
| ------- | ------------- | --------- | -------------------- | --------------------- | ---------------- | ----------------- | -------------------- |
| **MLP** | 0.8765        | 0.3428    | 235,146              | 2.72 MB               | ~2.4 M           | ~1.2 M            | ~50 MB               |
| **CNN** | 0.8909        | 0.3035    | 56,714               | 0.69 MB               | ~24 M            | ~12 M             | ~100 MB              |


### 📝 Task 5.2 : Analysis

 1- Accuracy:

CNN تفوّق بدقّة 0.8909 مقابل 0.8765 للـ MLP.

السبب أن الـ CNN يستغل العلاقات المكانية بين البكسلات ويستخرج ميزات محلية قوية (مثل الحواف والأنسجة).

2- Parameters & Model Size:

MLP يحتوي على معاملات أكثر (≈ 235 K مقابل ≈ 57 K).

ومع ذلك، حجمه على القرص أكبر (2.72 MB مقابل 0.69 MB).

هذا لأن الـ MLP يعتمد على طبقات كثيفة fully connected تتطلب وزنًا لكل اتصال.

3- Computational Complexity:

CNN يتطلب FLOPs أكثر أثناء التدريب بسبب عمليات الالتفاف convolution.

لكن هذه العمليات أكثر كفاءة من حيث استخراج الميزات، مما يؤدي لأداء أفضل حتى بموارد أكبر قليلًا.

4- Memory Usage:

MLP يستخدم ذاكرة أقل نسبيًا،

بينما CNN يحتاج لذاكرة إضافية لتخزين خرائط الميّزات feature maps أثناء التدريب.